src/bcc2.c

```diff
@@ -16,108 +17,126 @@
-        if (argc == 2) {
-          flags.list_platforms |= 1;
-        } else if (i + 1 >= argc || argv[i + 1][0] == '-') {
-          log_err_final("expected platform name after -platform");
-        } else {
-          flags.platform = NULL;
-          for (size_t j = 0; platforms[j] != NULL; j++) {
-            if (strcmp(platforms[j]->name, argv[i + 1]) == 0) {
-              flags.platform = platforms[j];
-              i++;
-              break;
-            }
-          }
-          if (flags.platform == NULL) {
-            log_err_final("unknown platform name '%s'", argv[i + 1]);
-          }
-        }
-      }
-    }
-  }
+print_help(const char *program_name, const char *program_description,
+           struct Option *opts[], size_t opts_size) {
+  printf("%s: %s\n", program_name, program_description);
+  printf("Usage: %s [OPTION]... [FILE]\n", program_name);
+  print_flags(opts, opts_size);
 }
 
+struct Option help = {
+    .flag = "h",
+    .description = "prints the help message",
+    .required_arg = ARG_NONE,
+    .type = OPT_BOOLEAN,
+    .long_flag = false,
+};
+struct Option version = {
+    .flag = "v",
+    .description = "prints the version",
+    .required_arg = ARG_NONE,
+    .type = OPT_BOOLEAN,
+    .long_flag = false,
+};
+struct Option ast_dump_flag = {
+    .flag = "dump-ast",
+    .description = "dumps the abstract syntax tree to stdout",
+    .required_arg = ARG_NONE,
+    .type = OPT_BOOLEAN,
+    .long_flag = true,
+};
+struct Option ir_dump_flag = {
+    .flag = "dump-ir",
+    .description = "dumps the intermediate representation to stdout",
+    .required_arg = ARG_NONE,
+    .type = OPT_BOOLEAN,
+    .long_flag = true,
+};
+struct Option reg_dump_flag = {
+    .flag = "dump-reg",
+    .description = "dumps the registers to stdout",
+    .required_arg = ARG_NONE,
+    .type = OPT_BOOLEAN,
+    .long_flag = true,
+};
+struct Option platform_flag = {
+    .flag = "platform",
+    .description = "selects the platform to compile for",
+    .argument_name = "platform_name",
+    .required_arg = ARG_OPTIONAL,
+    .type = OPT_STRING,
+    .long_flag = true,
+};
+struct Option list_platforms = {
+    .flag = "list-platforms",
+    .description = "lists all the platforms supported",
+    .required_arg = ARG_NONE,
+    .type = OPT_BOOLEAN,
+    .long_flag = true,
+};
+
 int
 main(int argc, char *argv[]) {
-  parse_args(argc, argv);
-
-  if (flags.help) {
-    printf("-h : prints help\n"
-           "-v : prints version\n"
-           "-ast : dumps ast to stdout\n"
-           "-ir : dumps ir to stdout\n"
-           "-regs : dumps registers to stdout\n"
-           "-platform : list platforms\n"
-           "-platform <platform> : selects the platform to compile for\n");
-    exit(EXIT_SUCCESS);
+  struct Option *opts[] = {&help,          &version,       &ast_dump_flag,
+                           &ir_dump_flag,  &reg_dump_flag, &platform_flag,
+                           &list_platforms};
```

Comment:
It may be worth it to make this array NULL terminated instead of computing its size with integer division.